src/core_plugins/kibana/public/visualize/embeddable/visualize_embeddable.js

```diff
@@ -93,7 +93,13 @@ export class VisualizeEmbeddable extends Embeddable  {
     }
   }
 
-  render(domNode) {
+  /**
+   *
+   * @param {Element} domNode
+   * @param {ContainerState} containerState
+   */
+  render(domNode, containerState) {
+    this.onContainerStateChanged(containerState);
     this.domNode = domNode;
```

Comment:
I'm a little concerned about the `this.onContainerStateChanged(containerState);` in the render. I know this isn't React, but in general, I think of render as a "Make the DOM look like this" kind of function. If we are doing `onContainerStateChanged` here, it seems like it might be easy to get into a loop where `onContainerStateChanged` makes a change, which then makes a render happen, which then calls `onContainerStateChanged`, etc. Just a thought!

Reply:
I think your concerns are legit but this code is going to change soon with some updates @timroes is making to the loader.  I think in the end it will be something like this pseudocode:

```
onContainerStateChanged(containerState) {
 this.saveState(containerState);
 this.reloadVisualization(getStateForRefresh());
}

render(dom, containerState) {
 this.saveState(containerState);
 this.createVisualization(domNode, getStateForFullRender());
}
```

Would that alleviate your concerns?  It would just refactor the actual storing of state into a separate function both things could call.